floodlight/models/kinematics.py

```diff
@@ -0,0 +1,102 @@
+import warnings
+from dataclasses import dataclass
+
+import numpy as np
+
+from floodlight.utils.types import Numeric
+from floodlight import Pitch, XY
+from floodlight.core.property import PlayerProperty
+
+
+@dataclass
+class ModelDistance:
+    """Class for calculating distances of players on the pitch.
+
+    Parameters
+    __________
+    pitch: Pitch
+        Pitch object of the data. Should at least include information about the data's
+        unit. If pitch.unit is 'percent', length and width of the pitch are also needed.
+
+    """
+
+    pitch: Pitch = None
+    length: Numeric = None
+    width: Numeric = None
+    unit: str = None
+
+    def __init__(self, pitch):
+        self.unit = pitch.unit
+        if pitch.length is not None:
+            self.length = pitch.length
+        if pitch.width is not None:
+            self.width = pitch.width
+        if pitch.unit is not None:
+            self.unit = pitch.unit
+        else:
+            warnings.warn(
+                "No information about pitch unit. Coordinates are assumed to be in m!"
+            )
+        self._distance_euclidean = None
+        self.framerate = None
+
+    def __str__(self):
+        return (
+            f"Floodlight ModelDistance Class with unit [{self.unit}],"
+            f"length {self.length} and width {self.width}"
+        )
+
+    def fit(self, xy: XY, direction: str = None):
+        """Fits a model to calculate euclidean distances to an xy object.
+
+        Parameters
+        ----------
+        xy: XY
+            Floodlight XY Data object.
+        direction: str
+            One of {'x', 'y', 'xy'} of None.
+            Indicates the direction to calculate the distance.
+            'x' will only calculate distance in the direction of the x-coordinate.
+            'y' will only calculate distance in the direction of the y-coordinate.
+            'xy' of None will calculate distance in both directions.
+
+        """
+
+        distance_euclidean = None
+
+        if self.unit is None or self.unit == "m":
+            unit_factor = (1, 1)
+        elif self.unit == "cm":
+            unit_factor = (100, 100)
+        elif self.unit == "percent":
+            unit_factor = (self.length / 100, self.width / 100)
+        else:
+            warnings.warn(
+                "No information about pitch unit. Coordinates are assumed to be in m!"
+            )
+            unit_factor = (1, 1)
+
+        if direction is None or direction == "xy":
+            differences_xy = np.gradient(xy.xy, axis=0)
+            distance_euclidean = np.hypot(
+                differences_xy[:, ::2] * unit_factor[0],
+                differences_xy[:, 1::2] * unit_factor[1],
+            )
+        elif direction == "x":
+            distance_euclidean = np.gradient(xy.x * unit_factor[0], axis=0)
+        elif direction == "y":
+            distance_euclidean = np.gradient(xy.y * unit_factor[1], axis=0)
```

Comment:
I'm not exactly sure what you are calculating here... for example, I've tried the following:

```
>>> xy = XY(np.array(((0, 0), (0, 1), (0, 2), (0, 3))))
>>> xy
XY(xy=array([[0, 0],
       [0, 1],
       [0, 2],
       [0, 3]]), framerate=None, direction=None)
>>> model.fit(xy)
>>> model.distance
PlayerProperty(property=array([[1.],
       [1.],
       [1.],
       [1.]]), name='distance', framerate=5)
```

So for a player that starts at (0,0) and walks 1 meter per frame, he get's 1 meter distance value per frame. But after 4 frames, he's walked 3 meters, but the (cumulative) distance would be 4. I assume this is the result of using the central difference in the np.gradient function, and of course it makes sense if in the next step you want to derive speeds or accelerations. But in itself, it's hard to interpret what this "distance" really means.

I'd suggest to include multiple measures (with informative naming), ideally in way they can be derived from performing only a single sweep through the array. For the example XY array above, this could  look something like this:
central_difference_distance = [1, 1, 1, 1]
distance_covered = [0, 1, 1, 1]
cumulative_distance_covered = [0, 1, 2, 3]

What do you think?

Reply:
I see where this is confusing.

The central difference method calculates the slope at t as the mean of the slopes of t-1 to t and t to t+1. Therefore, for the first and last frame, the data gets extrapolated by one frame. When applying the central difference method in biomechanics, you assume that the relevant event you want to capture does not begin at the first frame (and does not end at the last). In your example, I would argue, that if it is relevant for you that the player wasn't moving before, you should have started your observation one frame prior so you have  `xy = XY(np.array(((0, 0), (0, 0), (0, 1), ... )))´. Otherwise its fair to assume that he is moving and therefore moved 1 meter in the first frame as well.

I still think approaches to numerical differentiation (eg. forwards/backwards differentiation) can be included although the central difference method is the standard approach. However, I think "distance_covered" would be misleading in this regard. I would rather call them by their technical name and leave it to the users which one is applicable in their case. You would then get:
central_difference_distance = [1, 1, 1, 1]
backward_difference_distance = [0, 1, 1, 1]
forward_difference_distance = [1, 1, 1, 0]

I think cumulative distance should then be calculated based on the method of differentiation. So you would get:
cumulative_central_distance_covered = [1, 2, 3, 4]
cumulative_backward_distance_covered = [0, 1, 2, 3]
cumulative_forward_distance_covered = [1, 2, 3, 3]

Does this make sense?

Reply:
Yes, makes sense to me! Two follow-up questions result for me:

1) Is there a way to compute all three differentiation variants in one (which I doubt) step? I'm just thinking that if three differentiations would unavoidably mean three complete sweeps through the array this has some unwanted implications: Likely, when using this method in a given situation, I'm mostly interested in _one_ particular difference method, and computing all three to serve all use cases would be inefficient in this situation. There are three solutions that come to my mind:
- the difference method could be a parameter with "central" as the default, but this would mean I could subsequently only query the distance covered in this respect
- the differentiation could be taken out of the ``.fit()``-method and rather done on request when querying the distances, but this would make the ``.fit()``-method kind of redundant
- maybe discard the backward difference, as I don't see a use case for that straight away... could be a compromise resulting in all information possible desirable at the cost of two array sweeps

2) Although I agree with your explanation and accordingly the usage of technical terms, I'm thinking about the users who do not care so much about the details and "just want to see how far they ran". Having to call (if we go with the first suggestion of the previous point):
```
model.fit(difference="forward")
distances = model.cumulative_forward_distance_covered
```
is not really intuitive or event intimidating. Can you think of a compromise?

Reply:
As I see it, we could do it with two methods. I would go with your first suggestion and put the difference method as a parameter in the `.fit()`-method. Then we build a general `.cumulative_distance_covered()` method on top and calculate the cumulative distance on request. I don't see why the `np.cumsum()`-function would care about which difference method we used prior, so there is no need to build different methods for it. 
So the basic user would just call:
```
model.fit(xy, difference="central"/"forward")
distances = model.cumulative_distance_covered
```
Then, the `.fit()`-method does only what is says, i.e., fitting a specific model to the data. The cumulative distance arguably belongs inside the properties-class as we can calculate it for every given frame. But after that we move towards metrics pretty fast, as `.cumulative_distance[-1]` would already give us total distance covered. Do you already have a plan where to cut the line between those properties and metrics?

Reply:
Sounds good to me! I'd suggest du make two methods next to the ``.fit()``-method, ``distance_covered`` and ``cumulative_distance_covered`` and return ``PlayerProperty``s, respectively. This model is probably the easiest to infer a metric from the property, but I'd still leave that functionality to a different method as this will be the general case. Such a method then can also do "more advanced" calculations such as 5-minute-splits or else!

Reply:
To avoid confusion: I implemented the *backward* instead of the *forward* difference method as it will start with a 0 for the first frame.